docs/admin/validators.mdx

```diff
@@ -0,0 +1,95 @@
+---
+sidebar_position: 5
+sidebar_label: "Validator Commands"
+id: admin-validators
+title: Validator Information and Transactions
+description: Validator transactions (join/approve/leave) and queries with the validator command
+slug: /admin/validators
+---
+
+The `validators` command provides functions for creating and broadcasting
+validator-related transactions (join/approve/leave), and retrieving information
+on the current validators and join requests.
+
+```
+kwil-admin validators [--rpcserver RPCSERVER] <command> [<args>]
+```
+
+Each of the subcommands uses a Kwil node's public RPC service (`--rpcserver`),
+where the default is "127.0.0.1:50051".
```

Comment:
As such, I suspect we'll either see this command go away or become a node subcommand if it starts using the authenticated RPC service to perform server-side tx auth/signing.

Reply:
Kind've a weird one.

I think `join-status` should probably be on the public endpoint (however if this makes config weird / confusing by having to specify 2 endpoints, it's probably fine to include it with the rest).

`join` makes sense as a private authenticated call, however it is not clear what network it is joining (I am assuming that it is the network that it is already a non-validator of, but this feels a little bit weird to me).  Essentially, it needs the url of a node that it can forward the join request to (using broadcast).

`approve` and `leave` should definitely be using the authenticated service.

Maybe we can have a discussion about this tomorrow

Reply:
Yeah, we should all get on the same page about the validator actions.

These general M.O. of validator commands is unchanged since they were added when the first comet bft pr was merged to main.  That is, other than `list` and `join-status`:
- node key is on the machine running the command
- kwil-admin (previously kwild utils) authors and signs one of the three transactions (join, leave, approve)
- the transaction is broadcasted on _any public RPC service_

It's exactly because this is unchanged that these commands are not subcommands of the `node` command (yet).

That all makes sense if you think of if like a wallet, which has keys, makes transactions, and broadcasts.  I believe this model would be intuitive to many people.  This approach requires having the keys on the administrative machine.  It also allows using an arbitrary RPC endpoint to broadcast transactions on behalf of any number of nodes that the operator controls, without having to touch those node machines themselves.  These would not use an authenticated RPC service for broadcast.

On the other hand, we've discussed changing to a model where kwil-admin simply makes a request to the node (current or candidate validator) that should do the transaction authoring and signing with it's own in-memory keys.  The security model here puts all the responsibility on the RPC authentication.  Yet it keeps the node keys on the node.  When this model is implemented, the kwil-admin commands would probably go under the `node` command, which is described as the remote node administration commands, where all auth RPC is performed.  This is likely to be error prone since you may easily get the wrong endpoint for the node, and end up signing a transaction with the wrong keys (thus accidentally making your validator leave, or attempt to join, etc)

There are pros and cons to both approaches.

Reply:
The security concern of having to copy private keys across machines really irks me.  As far as I know, best security practices say to generate a key on a machine and not have it leave.  I think our default should be operating under this model.

The other option is that this admin tool could directly access the private key, and then validator sub commands can only be used when running this on the same machine.

I think either of these would be fine, but I do think it's important that the expected functionality is to run a validator without ever copying the key.

Reply:
I understand your point of view, but if you consider the Ethereum validator model, it's somewhat opposite.  Their official and recommended approach is to use a special key generation tool, `deposit`, then _copy_ the generated keys to the validator client machine, and import it into the validator client.

https://docs.prylabs.network/docs/install/install-with-script#step-5-run-a-validator-using-prysm
https://launchpad.ethereum.org/en/checklist
https://github.com/ethereum/staking-deposit-cli#new-mnemonic-arguments (has `--num_validators` option to generate multiple validator keys at once)

You are correct that the key go to only one validator machine, however that is to prevent two instances of the same validator client running, which leads to slashing.

While the copy is mainly necessary because the key generation tool is supposed to be done on a secure offline machine (and you have backups), there are also validator management tools that use the keys.

For example, Prysm has their own validator wallet to manage _multiple_ validator keys: https://docs.prylabs.network/docs/wallet/deterministic
Their approach to leave/exit is almost exactly like ours at present: https://docs.prylabs.network/docs/wallet/exiting-a-validator
`prysmctl validator exit --wallet-dir=<path/to/wallet> --beacon-rpc-provider=<127.0.0.1:4000> `
They have actually transitioned from a direct validator client command to this prysmctl approach that uses a wallet with validator keys and an RPC endpoint. ("Note that this operation was previously facilitated by a command exposed by the Prysm validator client")

We should definitely get kwild to make these validator transactions internally, but I'm not entirely convinced there's no use case for signing join/leave/approve transactions in an external tool like kwil-admin, even when it's on a different machine (likely to be a secure machine that contains multiple generated validator keys, as is the case with ethereum's `deposit` key store and prysmctl's wallet).